app/src/main/java/com/duckduckgo/app/cta/ui/CtaViewModel.kt

```diff
@@ -0,0 +1,193 @@
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class CtaViewModel @Inject constructor(
+    private val appInstallStore: AppInstallStore,
+    private val pixel: Pixel,
+    private val surveyDao: SurveyDao,
+    private val widgetCapabilities: WidgetCapabilities,
+    private val dismissedCtaDao: DismissedCtaDao,
+    private val variantManager: VariantManager
+) {
+
+    val ctaViewState: MutableLiveData<CtaViewState> = MutableLiveData()
+    val surveyLiveData: LiveData<Survey> = surveyDao.getLiveScheduled()
+    private var activeSurvey: Survey? = null
+
+    data class CtaViewState(
+        val cta: CtaConfiguration? = null
+    )
+
+    fun onSurveyChanged(survey: Survey?) {
+        activeSurvey = survey
+        refreshCta()
+    }
+
+    fun refreshCta() {
+        surveyCta()?.let {
+            ctaViewState.postValue(ctaViewState.value!!.copy(cta = it))
+            return
+        }
+
+        Schedulers.io().scheduleDirect {
+            if (canShowWidgetCta()) {
+                val ctaType = if (widgetCapabilities.supportsAutomaticWidgetAdd) AddWidgetAuto else AddWidgetInstructions
+                ctaViewState.postValue(ctaViewState.value!!.copy(cta = ctaType))
+            } else {
+                ctaViewState.postValue(ctaViewState.value!!.copy(cta = null))
+            }
+        }
+    }
+
+    private fun surveyCta(): CtaConfiguration.Survey? {
+        val survey = activeSurvey
+        if (survey?.url != null) {
+            val showOnDay = survey.daysInstalled?.toLong()
+            val daysInstalled = appInstallStore.daysInstalled()
+            if (showOnDay == null || showOnDay == daysInstalled) {
+                return Survey(survey)
+            }
+        }
+        return null
+    }
+
+    @WorkerThread
+    private fun canShowWidgetCta(): Boolean {
+        return variantManager.getVariant().hasFeature(AddWidgetCta) &&
+                widgetCapabilities.supportsStandardWidgetAdd &&
+                !widgetCapabilities.hasInstalledWidgets &&
+                !dismissedCtaDao.exists(CtaId.ADD_WIDGET)
+    }
+
+    fun onCtaShown() {
+        val cta = ctaViewState.value?.cta ?: return
+        pixel.fire(cta.shownPixel)
+    }
+
+    fun onCtaDismissed() {
+        val cta = ctaViewState.value?.cta ?: return
+        pixel.fire(cta.cancelPixel)
+
+        Schedulers.io().scheduleDirect {
+            when (cta) {
+                is CtaConfiguration.Survey -> {
+                    activeSurvey = null
+                    surveyDao.cancelScheduledSurveys()
+                }
+                else -> {
+                    dismissedCtaDao.insert(DismissedCta(cta.ctaId))
+                }
+            }
+            ctaViewState.postValue(ctaViewState.value?.copy(cta = null))
+            refreshCta()
+        }
+    }
+
+    fun onCtaLaunched() {
+        val cta = ctaViewState.value?.cta ?: return
+        pixel.fire(cta.okPixel)
+    }
+
+    init {
```

Comment:
`init` block at the bottom feels weird to me

Reply:
Same! I though it was in the official style guide but can't find a reference to it now. Will move it to the top, much nicer there.